{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "laughing-malaysia",
   "metadata": {},
   "source": [
    "Script to generate a list of columns in GEOME that have been required in at least one project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "czech-tract",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:loader:Loading: 114 - Collins Lab\n",
      "INFO:loader:Loading: 125 - European ARMS project\n",
      "INFO:loader:Loading: 131 - Amalda_mitochondria_genomes\n",
      "INFO:loader:Loading: 147 - Pillar Point - CALeDNA\n",
      "INFO:loader:Loading: 355 - NMNH TEST PROJECT\n",
      "INFO:loader:Loading: 183 - DiSSCo\n",
      "INFO:loader:Loading: 41 - Zostera capensis pooled RADseq\n",
      "INFO:loader:Loading: 332 - HIV PROJECT\n",
      "INFO:loader:Loading: 90 - Bionetwork\n",
      "INFO:loader:Loading: 61 - Smithsonian DROP Project\n",
      "INFO:loader:Loading: 67 - Cairns LAB\n",
      "INFO:loader:Loading: 83 - Windsor LAB\n",
      "INFO:loader:Loading: 376 - Early presence of Batrachochytrium dendrobatidis in Mexico with a contemporary dominance of the global panzootic lineage\n",
      "INFO:loader:Loading: 303 - Tata PowerPoint\n",
      "INFO:loader:Loading: 168 - Bighorn_RAD_Parentage_2018\n",
      "INFO:loader:Loading: 158 - Candida ITS\n",
      "INFO:loader:Loading: 181 - Mexican_gray_wolf_RAD_Parentage_2018\n",
      "INFO:loader:Loading: 89 - Mekong River\n",
      "INFO:loader:Loading: 118 - Crandall Lab\n",
      "INFO:loader:Loading: 13 - Genetic diversity of coral reef damselfish feeding guilds\n",
      "INFO:loader:Loading: 40 - Kuroshio Current\n",
      "INFO:loader:Loading: 68 - CMPI\n",
      "INFO:loader:Loading: 69 - Harasewych LAB\n",
      "INFO:loader:Loading: 70 - INDIAN RIVER LAGOON\n",
      "INFO:loader:Loading: 71 - Lemaitre LAB\n",
      "INFO:loader:Loading: 77 - Osborn LAB\n",
      "INFO:loader:Loading: 78 - PEER Deadheads\n",
      "INFO:loader:Loading: 154 - ITS\n",
      "INFO:loader:Loading: 201 - Australian intertidal fish\n",
      "INFO:loader:Loading: 191 - Walleye GT-seq -second trial\n",
      "INFO:loader:Loading: 284 - SNAPS: Student Network for Amphibian Pathogen Surveillance\n",
      "INFO:loader:Loading: 305 - Retrospective SRA Datathon\n",
      "INFO:loader:Loading: 75 - Moorea Biocode\n",
      "INFO:loader:Loading: 80 - PIRE ARMS\n",
      "INFO:loader:Loading: 309 - Global Silky Shark Pool-Seq Assessment\n",
      "INFO:loader:Loading: 81 - Strong LAB\n",
      "INFO:loader:Loading: 79 - Phillips LAB\n",
      "INFO:loader:Loading: 300 - Bd-Maps Legacy Project\n",
      "INFO:loader:Loading: 328 - Using multi-species seascape genomics to conserve areas of evolutionary importance\n",
      "INFO:loader:Loading: 190 - 1904 - Lake Erie walleye MSA\n",
      "INFO:loader:Loading: 291 - AmphibiaWeb Disease Portal\n",
      "INFO:loader:Loading: 220 - Apostle Islands\n",
      "INFO:loader:Loading: 348 - Population genomics of marine fisheries\n",
      "INFO:loader:Loading: 212 - Asymmetrical gene-flow in five co-distributed syngnathids explained by ocean currents and rafting propensity\n",
      "INFO:loader:Loading: 64 - MISE University of the Ryukyus\n",
      "INFO:loader:Loading: 76 - Norenburg LAB\n",
      "INFO:loader:Loading: 248 - Investigating the chytrid presence in a non-native amphibian in Cambridge, England\n",
      "INFO:loader:Loading: 293 - Gaither Lab\n",
      "INFO:loader:Loading: 319 - Atelopus patazensis\n",
      "INFO:loader:Loading: 316 - Sponging up Diversity\n",
      "INFO:loader:Loading: 242 - Kosnipata\n",
      "INFO:loader:Loading: 337 - Host-symbiont coevolution, cryptic structure, and bleaching susceptibility, in a coral species complex (Scleractinia; Poritidae).\n",
      "INFO:loader:Loading: 323 - Pfeiffer LAB\n",
      "INFO:loader:Loading: 72 - Meyer ARMS\n",
      "INFO:loader:Loading: 240 - Japumayo\n",
      "INFO:loader:Loading: 365 - A seasonal follow up of Batrachochytrium dendrobatidis (Bd) and amphibian species detection in Norwegian water- courses using environmental DNA\n",
      "INFO:loader:Loading: 366 - Bsal Negative Round Up\n",
      "INFO:loader:Loading: 295 - The Type Locality Project: collecting genomic-quality, topotypic vouchers and training the next generation of specimen-based researchers\n",
      "INFO:loader:Loading: 1 - Diversity of the IndoPacific\n",
      "INFO:loader:Loading: 221 - Widespread presence and high prevalence of Batrachochytrium dendrobatidis in Gabon\n",
      "INFO:loader:Loading: 290 - Fungal infection, decline, and persistence in the only obligate troglodytic Neotropical salamander\n",
      "INFO:loader:Loading: 317 - Identification and development of DNA barcodes on lobster (Panulirus spp.)\n",
      "INFO:loader:Loading: 361 - RESERVEBENEFIT\n",
      "INFO:loader:Loading: 344 - Tibetan Antelope_Microbiome_Hormone\n",
      "INFO:loader:Loading: 85 - SI Barcoding CBOL\n",
      "INFO:loader:Loading: 73 - Meyer Biocube\n",
      "INFO:loader:Loading: 327 - CALeDNA\n",
      "INFO:loader:Loading: 339 - Boardman Lake eDNA metabarcoding\n",
      "INFO:loader:Loading: 357 - CoGeDiv: Comparative Genomic Analysis of Species Divergence\n",
      "INFO:loader:Loading: 232 - FuTRES Sample Project\n",
      "INFO:loader:Loading: 82 - USAID INDO\n",
      "INFO:loader:Loading: 222 - Bd swabs from all Bd projects conducted by Dr Cori Richards Zawacki and collaborators\n",
      "INFO:loader:Loading: 282 - Florida Museum Environmental Archaeology Program Comparative Collection Data\n",
      "INFO:loader:Loading: 286 - Cryptic diversity of a widespread global pathogen reveals expanded threats to amphibian conservation\n",
      "INFO:loader:Loading: 74 - Meyer LAB\n",
      "INFO:loader:Loading: 223 - Batrachochytrium dendrobatidis invasion on California Islands\n",
      "INFO:loader:Loading: 243 - Survey of Pathogenic Chytrid Fungi (Batrachochytrium salamandrivorans) in Salamanders from Three Mountain Ranges in Europe and the Americas\n",
      "INFO:loader:Loading: 44 - Smithsonian BOEM Project\n",
      "INFO:loader:Loading: 255 - Batrachochytrium dendrobatidis in amphibian museum specimens from Costa Rica (UCR)\n",
      "INFO:loader:Loading: 254 - Sierra Nevada Retrospective Analysis\n",
      "INFO:loader:Loading: 241 - Is Chytridiomycosis an Emerging Infectious Disease in Asia?\n",
      "INFO:loader:Loading: 244 - Pathogen invasion and non-epizootic dynamics in Pacific newts in California over the last century\n",
      "INFO:loader:Loading: 354 - Bsal Consortium of Germany\n",
      "INFO:loader:Loading: 292 - USGS National Survey for the Amphibian Chytrid Fungus Batrachochytrium salamandrivorans\n"
     ]
    }
   ],
   "source": [
    "import logging\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import requests\n",
    "import concurrent.futures\n",
    "\n",
    "RULE_REQUIRED = \"RequiredValue\"\n",
    "ERROR_LEVEL = \"ERROR\"\n",
    "GEOME_API = \"https://api.geome-db.org/v1\"\n",
    "TIMEOUT = 30 #seconds\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "\n",
    "def processGeomeConfig(doc, proj_id=0, title=\"network\"):\n",
    "    res = []\n",
    "    defns = {}\n",
    "    for entity in doc.get(\"entities\", []):\n",
    "        alias = entity.get(\"conceptAlias\")\n",
    "        cols = set()\n",
    "        # Get the required columns\n",
    "        for r in entity.get(\"rules\"):\n",
    "            if r.get(\"name\") == RULE_REQUIRED and r.get(\"level\") == ERROR_LEVEL:\n",
    "                cols.update(r.get(\"columns\"))\n",
    "        # Get column definition info\n",
    "        for a in entity.get(\"attributes\", []):\n",
    "            cname = a.get(\"column\")\n",
    "            if cname in cols:\n",
    "                adef = {\n",
    "                    \"definedBy\": a.get(\"definedBy\"),\n",
    "                    \"group\": a.get(\"group\"),\n",
    "                    \"definition\": a.get(\"definition\")\n",
    "                }\n",
    "                defns[cname] = adef\n",
    "        for c in cols:\n",
    "            #row = (alias, c, proj_id, title)\n",
    "            row = (alias, c, proj_id)\n",
    "            res.append(row)\n",
    "    return res, defns\n",
    "\n",
    "def getProjectConfig(project_id):\n",
    "    url = f\"{GEOME_API}/projects/{project_id}/config\"\n",
    "    response = requests.get(url, timeout=TIMEOUT)\n",
    "    cfgdoc = response.json()\n",
    "    return cfgdoc\n",
    "\n",
    "\n",
    "def processProject(project_id, project_title):\n",
    "    doc = getProjectConfig(project_id)\n",
    "    res, defns = processGeomeConfig(doc, proj_id=project_id, title=project_title)\n",
    "    return res, defns\n",
    "\n",
    "def processProjects():\n",
    "    L = logging.getLogger(\"loader\")\n",
    "    url = f\"{GEOME_API}/projects/\"\n",
    "    params = {\"includePublic\": \"true\", \"admin\": \"false\"}\n",
    "    response = requests.get(url, params=params, timeout=TIMEOUT)\n",
    "    projects = response.json()\n",
    "    res = []\n",
    "    definitions = {}\n",
    "    with concurrent.futures.ThreadPoolExecutor(max_workers=20) as executor:\n",
    "        futures = []\n",
    "        for project in projects:\n",
    "            proj_id = project.get(\"projectId\")\n",
    "            title = project.get(\"projectTitle\")            \n",
    "            L.info(f\"Loading: {proj_id} - {title}\")\n",
    "            futures.append(executor.submit(processProject, proj_id, title))\n",
    "        for future in concurrent.futures.as_completed(futures):\n",
    "            reqs, defns = future.result()\n",
    "            res += reqs\n",
    "            definitions.update(defns)\n",
    "    return res, definitions\n",
    "\n",
    "\n",
    "required = []\n",
    "response = requests.get(f\"{GEOME_API}/network/1/config\", timeout=TIMEOUT)\n",
    "network_config = response.json()\n",
    "required, definitions = processGeomeConfig(network_config)\n",
    "preq, pdefns = processProjects()\n",
    "required += preq\n",
    "definitions.update(pdefns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "partial-tyler",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>record_alias</th>\n",
       "      <th>column_name</th>\n",
       "      <th>project_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Event</td>\n",
       "      <td>country</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Event</td>\n",
       "      <td>yearCollected</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Event</td>\n",
       "      <td>locality</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sample</td>\n",
       "      <td>eventID</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Diagnostics</td>\n",
       "      <td>materialSampleID</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1863</th>\n",
       "      <td>Diagnostics</td>\n",
       "      <td>materialSampleID</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1864</th>\n",
       "      <td>Diagnostics</td>\n",
       "      <td>diagnosticID</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1865</th>\n",
       "      <td>Diagnostics</td>\n",
       "      <td>diseaseDetected</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1866</th>\n",
       "      <td>Diagnostics</td>\n",
       "      <td>fatal</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1867</th>\n",
       "      <td>Diagnostics</td>\n",
       "      <td>diseaseTested</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1868 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     record_alias       column_name  project_id\n",
       "0           Event           country           0\n",
       "1           Event     yearCollected           0\n",
       "2           Event          locality           0\n",
       "3          Sample           eventID           0\n",
       "4     Diagnostics  materialSampleID           0\n",
       "...           ...               ...         ...\n",
       "1863  Diagnostics  materialSampleID         292\n",
       "1864  Diagnostics      diagnosticID         292\n",
       "1865  Diagnostics   diseaseDetected         292\n",
       "1866  Diagnostics             fatal         292\n",
       "1867  Diagnostics     diseaseTested         292\n",
       "\n",
       "[1868 rows x 3 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame.from_records(required, columns=[\"record_alias\", \"column_name\", \"project_id\"])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "accredited-receptor",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>record_alias</th>\n",
       "      <th>Diagnostics</th>\n",
       "      <th>Event</th>\n",
       "      <th>Event_Photo</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Sample_Photo</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>fastaSequence</th>\n",
       "      <th>fastqMetadata</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>column_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>basisOfRecord</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>collectorList</th>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coordinateUncertaintyInMeters</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>country</th>\n",
       "      <td>0</td>\n",
       "      <td>84</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>decimalLatitude</th>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>decimalLongitude</th>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>designDescription</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>diagnosticID</th>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>diseaseDetected</th>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>diseaseTested</th>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eventEnteredBy</th>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eventID</th>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>22</td>\n",
       "      <td>84</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fatal</th>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>filenames</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genus</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>georeferenceProtocol</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>habitat</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>institutionCode</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>instrumentModel</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kingdom</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>libraryLayout</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>librarySelection</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>librarySource</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>libraryStrategy</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locality</th>\n",
       "      <td>0</td>\n",
       "      <td>84</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>materialSampleID</th>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>22</td>\n",
       "      <td>56</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>measurementMethod</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>measurementType</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>measurementUnit</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>measurementValue</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>photoID</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phylum</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>platform</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>principalInvestigator</th>\n",
       "      <td>0</td>\n",
       "      <td>36</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleEnteredBy</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleType</th>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>samplingProtocol</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scientificName</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sequence</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>specificEpithet</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>subProject</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tissueID</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>40</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tissuePlate</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tissueWell</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yearCollected</th>\n",
       "      <td>0</td>\n",
       "      <td>84</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "record_alias                   Diagnostics  Event  Event_Photo  Sample  \\\n",
       "column_name                                                              \n",
       "basisOfRecord                            0      0            0      26   \n",
       "collectorList                            0     23            0       0   \n",
       "coordinateUncertaintyInMeters            0      1            0       0   \n",
       "country                                  0     84            0       0   \n",
       "decimalLatitude                          0     60            0       0   \n",
       "decimalLongitude                         0     60            0       0   \n",
       "designDescription                        0      0            0       0   \n",
       "diagnosticID                            28      0            0       0   \n",
       "diseaseDetected                         23      0            0       0   \n",
       "diseaseTested                           23      0            0       0   \n",
       "eventEnteredBy                           0     22            0       0   \n",
       "eventID                                  0     83           22      84   \n",
       "fatal                                   23      0            0       0   \n",
       "filenames                                0      0            0       0   \n",
       "genus                                    0      0            0      23   \n",
       "georeferenceProtocol                     0      1            0       0   \n",
       "habitat                                  0      1            0       0   \n",
       "identifier                               0      0            0       0   \n",
       "institutionCode                          0      0            0      25   \n",
       "instrumentModel                          0      0            0       0   \n",
       "kingdom                                  0      0            0       3   \n",
       "libraryLayout                            0      0            0       0   \n",
       "librarySelection                         0      0            0       0   \n",
       "librarySource                            0      0            0       0   \n",
       "libraryStrategy                          0      0            0       0   \n",
       "locality                                 0     84            0       0   \n",
       "materialSampleID                        28      0            0      83   \n",
       "measurementMethod                        3      0            0       0   \n",
       "measurementType                          3      0            0       0   \n",
       "measurementUnit                          3      0            0       0   \n",
       "measurementValue                         3      0            0       0   \n",
       "photoID                                  0      0           23       0   \n",
       "phylum                                   0      0            0      25   \n",
       "platform                                 0      0            0       0   \n",
       "principalInvestigator                    0     36            0       0   \n",
       "sampleEnteredBy                          0      0            0      23   \n",
       "sampleType                              23      0            0       0   \n",
       "samplingProtocol                         0      3            0       0   \n",
       "scientificName                           0      0            0       6   \n",
       "sequence                                 0      0            0       0   \n",
       "specificEpithet                          0      0            0      23   \n",
       "subProject                               0      0            0       1   \n",
       "tissueID                                 0      0            0       0   \n",
       "tissuePlate                              0      0            0       0   \n",
       "tissueWell                               0      0            0       0   \n",
       "yearCollected                            0     84            0       0   \n",
       "\n",
       "record_alias                   Sample_Photo  Tissue  fastaSequence  \\\n",
       "column_name                                                          \n",
       "basisOfRecord                             0       0              0   \n",
       "collectorList                             0       0              0   \n",
       "coordinateUncertaintyInMeters             0       0              0   \n",
       "country                                   0       0              0   \n",
       "decimalLatitude                           0       0              0   \n",
       "decimalLongitude                          0       0              0   \n",
       "designDescription                         0       0              0   \n",
       "diagnosticID                              0       0              0   \n",
       "diseaseDetected                           0       0              0   \n",
       "diseaseTested                             0       0              0   \n",
       "eventEnteredBy                            0       0              0   \n",
       "eventID                                   0       0              0   \n",
       "fatal                                     0       0              0   \n",
       "filenames                                 0       0              0   \n",
       "genus                                     0       0              0   \n",
       "georeferenceProtocol                      0       0              0   \n",
       "habitat                                   0       0              0   \n",
       "identifier                                0       0             41   \n",
       "institutionCode                           0       0              0   \n",
       "instrumentModel                           0       0              0   \n",
       "kingdom                                   0       0              0   \n",
       "libraryLayout                             0       0              0   \n",
       "librarySelection                          0       0              0   \n",
       "librarySource                             0       0              0   \n",
       "libraryStrategy                           0       0              0   \n",
       "locality                                  0       0              0   \n",
       "materialSampleID                         22      56              0   \n",
       "measurementMethod                         0       0              0   \n",
       "measurementType                           0       0              0   \n",
       "measurementUnit                           0       0              0   \n",
       "measurementValue                          0       0              0   \n",
       "photoID                                  23       0              0   \n",
       "phylum                                    0       0              0   \n",
       "platform                                  0       0              0   \n",
       "principalInvestigator                     0       0              0   \n",
       "sampleEnteredBy                           0       0              0   \n",
       "sampleType                                0       0              0   \n",
       "samplingProtocol                          0       0              0   \n",
       "scientificName                            0       0              0   \n",
       "sequence                                  0       0             41   \n",
       "specificEpithet                           0       0              0   \n",
       "subProject                                0       0              0   \n",
       "tissueID                                  0       1             40   \n",
       "tissuePlate                               0       3              0   \n",
       "tissueWell                                0       3              0   \n",
       "yearCollected                             0       0              0   \n",
       "\n",
       "record_alias                   fastqMetadata  \n",
       "column_name                                   \n",
       "basisOfRecord                              0  \n",
       "collectorList                              0  \n",
       "coordinateUncertaintyInMeters              0  \n",
       "country                                    0  \n",
       "decimalLatitude                            0  \n",
       "decimalLongitude                           0  \n",
       "designDescription                         57  \n",
       "diagnosticID                               0  \n",
       "diseaseDetected                            0  \n",
       "diseaseTested                              0  \n",
       "eventEnteredBy                             0  \n",
       "eventID                                    0  \n",
       "fatal                                      0  \n",
       "filenames                                 57  \n",
       "genus                                      0  \n",
       "georeferenceProtocol                       0  \n",
       "habitat                                    0  \n",
       "identifier                                57  \n",
       "institutionCode                            0  \n",
       "instrumentModel                           57  \n",
       "kingdom                                    0  \n",
       "libraryLayout                             57  \n",
       "librarySelection                          57  \n",
       "librarySource                             57  \n",
       "libraryStrategy                           57  \n",
       "locality                                   0  \n",
       "materialSampleID                           0  \n",
       "measurementMethod                          0  \n",
       "measurementType                            0  \n",
       "measurementUnit                            0  \n",
       "measurementValue                           0  \n",
       "photoID                                    0  \n",
       "phylum                                     0  \n",
       "platform                                  57  \n",
       "principalInvestigator                      0  \n",
       "sampleEnteredBy                            0  \n",
       "sampleType                                 0  \n",
       "samplingProtocol                           0  \n",
       "scientificName                             0  \n",
       "sequence                                   0  \n",
       "specificEpithet                            0  \n",
       "subProject                                 0  \n",
       "tissueID                                  56  \n",
       "tissuePlate                                0  \n",
       "tissueWell                                 0  \n",
       "yearCollected                              0  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfg = df.groupby(by=[\"record_alias\",\"column_name\"]).count()\n",
    "pd.options.display.float_format = '{:,.0f}'.format\n",
    "p = pd.pivot_table(df, values=\"project_id\", index=\"column_name\", columns=\"record_alias\", \n",
    "                   aggfunc=lambda x: len(x)).fillna(0)\n",
    "p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "central-bishop",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "country\n",
      "       group: \n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/country\n",
      "  definition: The name of the country or major administrative unit in which the\n",
      "              Location occurs.\n",
      "\n",
      "locality\n",
      "       group: \n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/locality\n",
      "  definition: The specific description of the place. Less specific geographic\n",
      "              information can be provided in other geographic terms\n",
      "              (higherGeography, continent, country, stateProvince,\n",
      "              county, municipality, waterBody, island, islandGroup).\n",
      "              This term may contain information modified from the\n",
      "              original to correct perceived errors or standardize the\n",
      "              description.\n",
      "\n",
      "yearCollected\n",
      "       group: \n",
      "   definedBy: urn:yearCollected\n",
      "  definition: The year the collecting event took place\n",
      "\n",
      "eventID\n",
      "       group: Record and Owner Details\n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/eventID\n",
      "  definition: The collector's event identifier.\n",
      "\n",
      "diagnosticID\n",
      "       group: \n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/measurementID\n",
      "  definition: The identifier for the diagnostic procedure.\n",
      "\n",
      "materialSampleID\n",
      "       group: \n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/MaterialSampleID\n",
      "  definition: The collector's specimen number. This number must be unique among the\n",
      "              IDs within the expedition.\n",
      "\n",
      "photoID\n",
      "       group: \n",
      "   definedBy: \n",
      "  definition: identifier for the photo\n",
      "\n",
      "sequence\n",
      "       group: \n",
      "   definedBy: \n",
      "  definition: Sequence data for the material sample.\n",
      "\n",
      "identifier\n",
      "       group: \n",
      "   definedBy: \n",
      "  definition: \n",
      "\n",
      "libraryLayout\n",
      "       group: \n",
      "   definedBy: \n",
      "  definition: \n",
      "\n",
      "libraryStrategy\n",
      "       group: \n",
      "   definedBy: \n",
      "  definition: \n",
      "\n",
      "librarySource\n",
      "       group: \n",
      "   definedBy: \n",
      "  definition: \n",
      "\n",
      "librarySelection\n",
      "       group: \n",
      "   definedBy: \n",
      "  definition: \n",
      "\n",
      "platform\n",
      "       group: \n",
      "   definedBy: \n",
      "  definition: \n",
      "\n",
      "instrumentModel\n",
      "       group: \n",
      "   definedBy: \n",
      "  definition: \n",
      "\n",
      "designDescription\n",
      "       group: \n",
      "   definedBy: \n",
      "  definition: \n",
      "\n",
      "filenames\n",
      "       group: \n",
      "   definedBy: \n",
      "  definition: List of filenames to be uploaded to NCBI Genbank SRA\n",
      "\n",
      "institutionCode\n",
      "       group: Protocol and Storage Details\n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/ownerInstitutionCode\n",
      "  definition: The name (or acronym) in use by the institution having custody of the\n",
      "              object(s) or information referred to in the record.\n",
      "\n",
      "sampleEnteredBy\n",
      "       group: Record and Owner Details\n",
      "   definedBy: urn:EnteredBy\n",
      "  definition: The name of the person that entered the data.\n",
      "\n",
      "phylum\n",
      "       group: Taxonomy and Life History\n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/phylum\n",
      "  definition: The full scientific name of the phylum in which the taxon is\n",
      "              classified\n",
      "\n",
      "decimalLatitude\n",
      "       group: \n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/decimalLatitude\n",
      "  definition: The geographic latitude (in decimal degrees, using the spatial\n",
      "              reference system given in geodeticDatum) of the\n",
      "              geographic center of a Location. Positive values are\n",
      "              north of the Equator, negative values are south of it.\n",
      "              Legal values lie between -90 and 90, inclusive.\n",
      "\n",
      "decimalLongitude\n",
      "       group: \n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/decimalLongitude\n",
      "  definition: The geographic longitude (in decimal degrees, using the spatial\n",
      "              reference system given in geodeticDatum) of the\n",
      "              geographic center of a Location. Positive values are\n",
      "              east of the Greenwich Meridian, negative values are west\n",
      "              of it. Legal values lie between -180 and 180, inclusive.\n",
      "\n",
      "eventEnteredBy\n",
      "       group: \n",
      "   definedBy: urn:enteredBy\n",
      "  definition: The name of the person entering the data.\n",
      "\n",
      "tissueID\n",
      "       group: \n",
      "   definedBy: urn:tissueID\n",
      "  definition: \n",
      "\n",
      "kingdom\n",
      "       group: Taxonomy and Life History\n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/kingdom\n",
      "  definition: The full scientific name of the kingdom in which the taxon is\n",
      "              classified.\n",
      "\n",
      "scientificName\n",
      "       group: Diagnostics and Traits\n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/scientificName\n",
      "  definition: scientificName as the lowest taxon identified\n",
      "\n",
      "tissuePlate\n",
      "       group: \n",
      "   definedBy: urn:plateID\n",
      "  definition: The name of the plate.  Typically this is a 96 well plat for Biocode\n",
      "              projects\n",
      "\n",
      "tissueWell\n",
      "       group: \n",
      "   definedBy: urn:wellID\n",
      "  definition: The well location in the plate\n",
      "\n",
      "principalInvestigator\n",
      "       group: \n",
      "   definedBy: urn:principalInvestigator\n",
      "  definition: Firstname Lastname (the PI on the grant that generated this data)\n",
      "\n",
      "collectorList\n",
      "       group: \n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/recordedBy\n",
      "  definition: A list (concatenated and separated) of names of people, groups, or\n",
      "              organizations responsible for recording the original\n",
      "              Occurrence. The primary collector or observer,\n",
      "              especially one who applies a personal identifier\n",
      "              (recordNumber), should be listed first.  Name separation\n",
      "              should use the pipe ('|') symbol\n",
      "\n",
      "basisOfRecord\n",
      "       group: Record and Owner Details\n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/basisOfRecord\n",
      "  definition: The specific nature of the data record.  Examples include\n",
      "              PreservedSpecimen, FossilSpecimen, LivingSpecimen,\n",
      "              MaterialSample, Event, HumanObservation,\n",
      "              MachineObservation, Taxon, Occurrence\n",
      "\n",
      "genus\n",
      "       group: Taxonomy and Life History\n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/genus\n",
      "  definition: The full scientific name of the genus in which the taxon is\n",
      "              classified. This should be a text string starting with a\n",
      "              capital letter. If Unknown, then add note of its next\n",
      "              closest taxonomic level of identification in\n",
      "              taxonRemarks\n",
      "\n",
      "specificEpithet\n",
      "       group: Taxonomy and Life History\n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/specificEpithet\n",
      "  definition: The full scientific name of the specificEpithet in which the taxon is\n",
      "              classified. This should be a text string in lower case.\n",
      "              If unknown, use \"sp.\" (without quotes).\n",
      "\n",
      "sampleType\n",
      "       group: Diagnostics and Traits\n",
      "   definedBy: \n",
      "  definition: Type of sample tested: eg, swabbing, blood, skin, liver\n",
      "\n",
      "diseaseTested\n",
      "       group: Diagnostics and Traits\n",
      "   definedBy: \n",
      "  definition: Controlled vocabulary: Bd, Bsal, Bd+Bsal, Other\n",
      "\n",
      "diseaseDetected\n",
      "       group: Diagnostics and Traits\n",
      "   definedBy: \n",
      "  definition: Result of disease testing: True=Positive, False=Negative, No\n",
      "              Confidence=Ambiguous\n",
      "\n",
      "fatal\n",
      "       group: Diagnostics and Traits\n",
      "   definedBy: \n",
      "  definition: At time of sampling, presumed cause of death to be diseaseTested, mark\n",
      "              [FALSE] if unknown\n",
      "\n",
      "samplingProtocol\n",
      "       group: \n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/samplingProtocol\n",
      "  definition: The name of, reference to, or description of the method or protocol\n",
      "              used in acquiring the materialSample.\n",
      "\n",
      "measurementMethod\n",
      "       group: Diagnostics and Traits\n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/measurementMethod\n",
      "  definition: A description of or reference to (publication, URI) the method or\n",
      "              protocol used to determine the measurement, fact,\n",
      "              characteristic, or assertion.\n",
      "\n",
      "measurementUnit\n",
      "       group: Diagnostics and Traits\n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/measurementUnit\n",
      "  definition: The units associated with the measurementValue.\n",
      "\n",
      "measurementType\n",
      "       group: Diagnostics and Traits\n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/measurementType\n",
      "  definition: The nature of the measurement, fact, characteristic, or assertion.\n",
      "\n",
      "measurementValue\n",
      "       group: Diagnostics and Traits\n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/measurementValue\n",
      "  definition: The value of the measurement, fact, characteristic, or assertion.\n",
      "\n",
      "coordinateUncertaintyInMeters\n",
      "       group: \n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/coordinateUncertaintyInMeters\n",
      "  definition: The horizontal distance (in meters) from the given decimalLatitude and\n",
      "              decimalLongitude describing the smallest circle\n",
      "              containing the whole of the Location. Leave the value\n",
      "              empty if the uncertainty is unknown, cannot be\n",
      "              estimated, or is not applicable (because there are no\n",
      "              coordinates). Zero is not a valid value for this term.\n",
      "\n",
      "georeferenceProtocol\n",
      "       group: \n",
      "   definedBy: http://rs.tdwg.org/dwc/terms/georeferenceProtocol\n",
      "  definition: A description or reference to the methods used to determine the\n",
      "              spatial footprint, coordinates, and uncertainties (e.g.\n",
      "              GPS, Approximated from Google Maps, Based on locality\n",
      "              name search in Google Maps (Ben Morris script) etc.)\n",
      "\n",
      "habitat\n",
      "       group: \n",
      "   definedBy: http://purl.obolibrary.org/obo/ENVO_01000739\n",
      "  definition: A category or description of the habitat in which the Event occurred.\n",
      "\n",
      "subProject\n",
      "       group: Record and Owner Details\n",
      "   definedBy: urn:subProject\n",
      "  definition: A designation of a child projecct within the larger project.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import textwrap\n",
    "for k,d in definitions.items():\n",
    "    print(f\"{k}\")\n",
    "    print(f\"       group: {d.get('group') or ''}\")\n",
    "    print(f\"   definedBy: {d.get('definedBy') or ''}\")\n",
    "    print(f\"  definition: {textwrap.fill(d.get('definition') or '', 70, subsequent_indent='              ')}\")\n",
    "    print(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "angry-effect",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
